File: frontend/src/shared/api/projects.ts
import type {
  Project,
  ProjectCreateForLeadPayload,
  ProjectCreatePayload,
  ProjectPriority,
  ProjectStatus,
  ProjectUpdatePayload,
} from "@/entities/project/types";
import { apiRequest } from "@/shared/api/http";

export function fetchProjects(params?: {
  status?: ProjectStatus;
  lead_id?: number;
  is_active?: boolean;
  include_inactive?: boolean;
  limit?: number;
  offset?: number;
}) {
  const q = new URLSearchParams();
  if (params?.status) q.set("status", params.status);
  if (params?.lead_id != null) q.set("lead_id", String(params.lead_id));
  if (params?.is_active != null) q.set("is_active", String(params.is_active));
  if (params?.include_inactive) q.set("include_inactive", "true");
  if (params?.limit != null) q.set("limit", String(params.limit));
  if (params?.offset != null) q.set("offset", String(params.offset));
  const qs = q.toString();
  return apiRequest<Project[]>(`/projects${qs ? `?${qs}` : ""}`);
}

export function fetchProjectsActive() {
  return apiRequest<Project[]>(`/projects/active`);
}

export function fetchProjectsOverdue() {
  return apiRequest<Project[]>(`/projects/overdue`);
}

export function fetchProject(id: number) {
  return apiRequest<Project>(`/projects/${id}`);
}

export function createProject(body: ProjectCreatePayload) {
  return apiRequest<Project>("/projects", { method: "POST", json: body });
}

export function createProjectForLead(leadId: number, body: ProjectCreateForLeadPayload) {
  return apiRequest<Project>(`/leads/${leadId}/projects`, { method: "POST", json: body });
}

export function updateProject(id: number, patch: ProjectUpdatePayload) {
  return apiRequest<Project>(`/projects/${id}`, { method: "PATCH", json: patch });
}

export function deleteProject(id: number) {
  return apiRequest<void>(`/projects/${id}`, { method: "DELETE" });
}

export function fetchLeadProjects(leadId: number, include_inactive = false) {
  const q = include_inactive ? "?include_inactive=true" : "";
  return apiRequest<Project[]>(`/leads/${leadId}/projects${q}`);
}

export type { ProjectPriority, ProjectStatus };

File: app/api/v1/projects.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_db
from app.models import ProjectStatus
from app.schemas.project import ProjectCreate, ProjectListItem, ProjectRead, ProjectUpdate
from app.services.project import (
    ProjectNotFoundError,
    ProjectService,
    ProjectStateError,
)

router = APIRouter(prefix="/projects", tags=["projects"])


def _to_list_item(p) -> ProjectListItem:
    return ProjectListItem(
        id=p.id,
        lead_id=p.lead_id,
        title=p.title,
        status=p.status,
        priority=p.priority,
        deadline=p.deadline,
        is_active=p.is_active,
        created_at=p.created_at,
    )


@router.get("/active", response_model=list[ProjectListItem])
async def list_active_projects(
    db: Annotated[AsyncSession, Depends(get_db)],
) -> list[ProjectListItem]:
    rows = await ProjectService(db).list_active()
    return [_to_list_item(p) for p in rows]


@router.get("/overdue", response_model=list[ProjectListItem])
async def list_overdue_projects(
    db: Annotated[AsyncSession, Depends(get_db)],
) -> list[ProjectListItem]:
    rows = await ProjectService(db).list_overdue()
    return [_to_list_item(p) for p in rows]


@router.get("", response_model=list[ProjectRead])
async def list_projects(
    db: Annotated[AsyncSession, Depends(get_db)],
    status_filter: ProjectStatus | None = Query(None, alias="status"),
    lead_id: int | None = None,
    is_active: bool | None = None,
    include_inactive: bool = Query(
        False,
        description="Если true — не ограничивать по is_active, пока не задан явный is_active",
    ),
    limit: int = Query(200, ge=1, le=500),
    offset: int = Query(0, ge=0),
) -> list[ProjectRead]:
    rows = await ProjectService(db).list_projects(
        status=status_filter,
        lead_id=lead_id,
        is_active=is_active,
        include_inactive=include_inactive,
        limit=limit,
        offset=offset,
    )
    return [ProjectRead.model_validate(p) for p in rows]


@router.post("", response_model=ProjectRead, status_code=status.HTTP_201_CREATED)
async def create_project(
    body: ProjectCreate,
    db: Annotated[AsyncSession, Depends(get_db)],
) -> ProjectRead:
    try:
        p = await ProjectService(db).create(body)
    except ProjectNotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e) or "Lead not found")
    return ProjectRead.model_validate(p)


@router.get("/{project_id}", response_model=ProjectRead)
async def get_project(
    project_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
) -> ProjectRead:
    p = await ProjectService(db).get(project_id)
    if p is None:
        raise HTTPException(status_code=404, detail="Project not found")
    return ProjectRead.model_validate(p)


@router.patch("/{project_id}", response_model=ProjectRead)
async def update_project(
    project_id: int,
    body: ProjectUpdate,
    db: Annotated[AsyncSession, Depends(get_db)],
) -> ProjectRead:
    try:
        p = await ProjectService(db).update(project_id, body)
    except ProjectStateError as e:
        raise HTTPException(status_code=409, detail=str(e))
    if p is None:
        raise HTTPException(status_code=404, detail="Project not found")
    return ProjectRead.model_validate(p)


@router.delete("/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_project(
    project_id: int,
    db: Annotated[AsyncSession, Depends(get_db)],
) -> None:
    ok = await ProjectService(db).soft_delete(project_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Project not found")
